java-api/src/java9/java/xyz/wagyourtail/jvmdg/j9/intl/module/ModuleConstantHelper.java

```diff
@@ -0,0 +1,78 @@
+package xyz.wagyourtail.jvmdg.j9.intl.module;
+
+import xyz.wagyourtail.jvmdg.j9.stub.java_base.J_L_M_ModuleDescriptor;
+import xyz.wagyourtail.jvmdg.j9.stub.java_base.J_L_Module;
+import xyz.wagyourtail.jvmdg.j9.stub.java_base.J_L_ModuleLayer;
+
+import java.util.*;
+
+public class ModuleConstantHelper {
+    private static final Map<String, J_L_Module> MODULES = new HashMap<>();
+    private static final Map<String, J_L_Module> MODULES_FROM_PACKAGE = new HashMap<>();
+    private static final ClassLoader ORIGINAL_SYSTEM_CLASS_LOADER = ClassLoader.getSystemClassLoader();
+    public static final J_L_ModuleLayer BOOT_LAYER =
+            // Note: This field is in this class as it's initialization is execution order sensitive as it need
+            // to be loaded after "MODULES" HashMap is initialized, but before initializing any boot "Module"
+            new J_L_ModuleLayer(Collections.emptyList(), MODULES);
+
+    static {
+        // TODO: Auto generate this data in a separate "ModuleConstants" class
+        registerBootModule("java.base", new String[]{"java.io", "java.lang", "java.math",
+                "java.net", "java.nio", "java.security", "java.text", "java.time", "java.util",
+                "javax.crypto", "javax.net", "javax.security"});
+        registerBootModule("java.logging", new String[]{"java.util.logging"});
+        registerBootModule("java.net.http", new String[]{"java.net.http"});
+        registerBootModule("jdk.httpserver", new String[]{"com.sun.net.httpserver"});
+        registerBootModule("jdk.jfr", new String[]{"jdk.jfr"});
+    }
+
+    private static void registerBootModule(String name, String[] packages) {
+        J_L_M_ModuleDescriptor descriptor = new J_L_M_ModuleDescriptor.Builder(name, false, Collections.emptySet())
+                .packages(new HashSet<>(Arrays.asList(packages))).build();
+        J_L_Module module = new J_L_Module(ORIGINAL_SYSTEM_CLASS_LOADER, BOOT_LAYER, descriptor);
+        MODULES.put(name, module);
+        for (String packageName : packages) {
+            MODULES_FROM_PACKAGE.put(packageName, module);
+        }
+    }
+
+    public static J_L_Module bootModuleFromClass(Class<?> clazz) {
+        boolean bootClassLoader;
+        if ((bootClassLoader = clazz.getClassLoader() == null) ||
+                clazz.getClassLoader() == ORIGINAL_SYSTEM_CLASS_LOADER ||
+                clazz.getName().startsWith("xyz.wagyourtail.jvmdg.")) {
```

Comment:
string class name comparisons won't work well when api is shaded and relocated.
`contains` would probably work tho... it's unlikely that will cause issues

should jvmdg classes even be treated special tho. they won't be when an application is running on java 9+

Reply:
The check is needed to properly register JvmDowngrader stub modules classes in their respective modules.

I can compute the package name dynamically instead of hardcoding the package name, to make sure repackaging is supported though.

If jvmdg check were dropped, it would just cause jvmdg stubs to appear as unnamed module instead of the module we are trying to stub, so not really a big deal.

Now that I'm thinking about it, maybe it's better to hook `getModule` in a global hook instead, I'll remove the jvmdg check for now.

Reply:
Applied your suggestion